Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm

```diff
@@ -34,21 +34,45 @@
 
 #import "WKWebViewInternal.h"
 #import "WebExtensionController.h"
+#import "WebPageProxy.h"
+#import "WebProcessProxy.h"
 #import "_WKWebExtensionControllerDelegatePrivate.h"
 #import "_WKWebExtensionControllerInternal.h"
 
 namespace WebKit {
 
 void WebExtensionContext::addListener(WebPageProxyIdentifier identifier, WebExtensionEventListenerType type)
 {
+    auto page = WebProcessProxy::webPage(identifier);
+    if (!page)
+        return;
+
     if (!extension().backgroundContentIsPersistent() && m_backgroundWebView.get()._page->identifier() == identifier)
-        m_backgroundPageListeners.add(type);
+        m_backgroundContentEventListeners.add(type);
+
+    auto result = m_eventListenerPages.add(type, WeakPageCountedSet { });
+    result.iterator->value.add(*page);
```

Comment:
Do we need something like this logic for WebNavigation events?

Reply:
Or will those get filtered through to here?

Reply:
Nevermind, this is probably handled in response to `Messages::WebExtensionContext::AddListener`, which we call from the WebNavigation events.